openedx/core/djangoapps/xblock/api.py

```diff
@@ -38,15 +39,24 @@ def get_runtime_system():
     keep application startup faster, it's only initialized when first accessed
     via this method.
     """
-    # pylint: disable=protected-access
-    if not hasattr(get_runtime_system, '_system'):
+    # The runtime system should not be shared among threads, as there is currently a race condition when parsing XML
+    # that can lead to duplicate children.
+    # (In BlockstoreXBlockRuntime.get_block(), has_cached_definition(def_id) returns false so parse_xml is called, but
+    # meanwhile another thread parses the XML and caches the definition; then when parse_xml gets to XML nodes for
+    # child blocks, it appends them to the children already cached by the other thread and saves the doubled list of
+    # children; this happens only occasionally but is very difficult to avoid in a clean way due to the API of parse_xml
+    # and XBlock field data in general [does not distinguish between setting initial values during parsing and changing
+    # values at runtime due to user interaction], and how it interacts with BlockstoreFieldData. Keeping the caches
+    # local to each thread completely avoids this problem.)
+    cache_name = '_system_{}'.format(threading.get_ident())
```

Comment:
Do we know how this data is cleaned up and recycled (if ever)?

Reply:
This stores a `XBlockRuntimeSystem` for each thread, and it won't get cleaned up, but it's not a large data structure as far as I'm aware. It does in turn initialize a `BlockstoreFieldData` class which has its own cache that uses up memory, but that cache [automatically frees its own data up as needed](https://github.com/edx/edx-platform/blob/09f4d734ddd951a8b90460075991d510ecd8712e/openedx/core/djangoapps/xblock/runtime/blockstore_field_data.py#L247-L260).